FIX allow to output error_score if a scoring failure happen

dsandeep0138 opened the pull request:
- If a scorer fails during the model selection, set score to error_score similar to the fit failing scenario
- Modify scorer fail test cases to expect warning messages instead of a ValueError exception
- Add a new test case to test score failures.

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Fixes #17589 
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
BaseSearchCV _fit_and_score should move scoring into try/except as some scorers can raise an error for particular CV splits

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

dsandeep0138 committed:
Set score to error_score when a scorer fails

- If a scorer fails during the model selection, set score to
  error_score similar to the fit failing scenario
- Modify scorer fail test cases to expect warning messages
  instead of a ValueError exception

dsandeep97 committed:
Set score to error_score when a scorer fails

- If a scorer fails during the model selection, set score to
  error_score similar to the fit failing scenario
- Modify scorer fail test cases to expect warning messages
  instead of a ValueError exception

amueller commented:
Thanks! What's the motivation for 0 instead of using ``error_score``? It seems a bit dangerous.

dsandeep0138 commented:
Hey @amueller, sorry for that, it was supposed to be error_score (basically, the same logic as fit fail). I'll modify the commit message to reflect the same. Can you please let me know if there are any other comments?

adrinjalali commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -465,12 +468,13 @@ def test_raises_on_score_list():
     clf = DecisionTreeClassifier()
-    with pytest.raises(ValueError):
-        cross_val_score(clf, X, y, scoring=f1_scorer_no_average)
+    assert_warns(FitFailedWarning,
```
we prefer `pytest.raises` construct to `assert_warns` and `assert_raises` in tests.

adrinjalali replied:
Isn't the `FitFailedWarning` a bit misleading here? (unless I'm misunderstanding this)

dsandeep0138 replied:
@adrinjalali Thanks for the comments. With the new changes, the scorer fail uses the same try/exception block as the fit function to update the score in case of failure. 

https://github.com/scikit-learn/scikit-learn/blob/77ccf19c1f5c0fd756e906e098e4ceadc964663b/sklearn/model_selection/_validation.py#L533-L552

Hence we are getting the FitFailedWarning. Do you want me to modify the name to some generic name? If so, can you suggest some?

With these changes, the test cases no longer throw ValueError, so I had a discussion in the linked issue with @jnothman and he suggested to check for the fit failed warning instead. Can you please let me know how to proceed with this? Thanks.

adrinjalali replied:
I don't think we can/should change the name of the warning for when `fit` fails cause that'll be backward incompatible. Ideally we'd use either a new `ScorinFailedWarning` or just a `UserWarning` or something for when the scoring fails, i.e. in two different `try ... catch` blocks.

To test for warnings, you can do

``` python
    with warnings.catch_warnings(record=True) as warn:
        ...
    assert warn
```
to make sure no warning is raised, or to check `warn` and make sure the right warning is raised among them.

There are a few examples in the tests which you should be able to find if you look for `with warnings.catch_warnings(record=True) as`

dsandeep0138 replied:
Done

dsandeep0138 replied:
I have used ``UserWarning`` with a separate try exception block.

adrinjalali reviewed (commented):
Thanks for the PR @dsandeep0138

dsandeep0138 committed:
Modify tests to use catch_warnings.

- Tests are mofified to catch warnings instead of assert_warns
- UserWarning is thrown now when scoring fails, instead of reusing
  the FitFailedWarning which is misleading

dsandeep0138 committed:
Fix lint issues

dsandeep0138 committed:
Fix lint issues#2

dsandeep0138 committed:
Make test case better, identifying the ValueError in error message.

dsandeep0138 committed:
Update the comment.

dsandeep0138 committed:
Modify all the test cases to use catch_warnings

dsandeep0138 committed:
Fix some of the remaining lint issues.

dsandeep0138 committed:
Include more tests to improve code coverage

dsandeep0138 committed:
changes to improve patch test coverage.

dsandeep0138 committed:
Improvise test cases; Add new test case for score failures.

adrinjalali reviewed (commented):
I think the solution now makes more sense.

dsandeep0138 commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -459,18 +460,43 @@ def test_supervised_cluster_scorers():
+                       "Details: \n%s" % (score_kwargs['error_score'],
+                                          error_message))
+
+    with warnings.catch_warnings(record=True) as record:
+        warnings.simplefilter("always")
         cross_val_score(clf, X, y, scoring=f1_scorer_no_average)
+        assert len(record) > 0
+        for item in record:
+            assert 'Traceback (most recent call last):\n' in str(item.message)
+            split = str(item.message).splitlines()
+            mtb = split[0] + '\n' + split[-1]
+            assert warning_message in mtb
```
I think the catch_warnings is useful, since I'm able to verify the ValueError as well in the traceback, not just check if a certain warning is raised. Please let me know what do you think. Please let me know if you have any comments on the patch and does it need any modifications?

dsandeep0138 committed:
Use pytest.warns instead of warnings.catch_warnings

dsandeep0138 committed:
Resolve lint issues

dsandeep0138 commented:
@adrinjalali I have modified the patch to use ``pytest.warns``, can you please review and let me know your comments?

adrinjalali commented on sklearn/metrics/tests/test_score_objects.py:
```diff
@@ -459,18 +459,38 @@ def test_supervised_cluster_scorers():
+
+    with pytest.warns(UserWarning, match=warning_message) as record:
         cross_val_score(clf, X, y, scoring=f1_scorer_no_average)
+    assert len(record) > 0
+    for item in record:
+        assert 'Traceback (most recent call last):\n' in str(item.message)
+        split = str(item.message).splitlines()
+        assert error_message in split[-1]
```
i think we usually test for this, I feel like it's too strict of a test

dsandeep0138 replied:
@adrinjalali Removed and simplified, can you please review?

dsandeep0138 committed:
Simplify test case.

dsandeep0138 committed:
Fix lint issues

adrinjalali commented on sklearn/model_selection/_validation.py:
```diff
@@ -573,13 +573,31 @@ def _fit_and_score(estimator, X, y, scorer, train, test, verbose,
+                raise
+            elif isinstance(error_score, numbers.Number):
+                score_time = time.time() - start_time - fit_time
+                test_scores = {name: error_score for name in scorer}
+                if return_train_score:
+                    train_scores = test_scores.copy()
```
should this logic move to `_score` and add an extra parameter to `_score` to handle different `error_score` cases? Also, when only one scorer fails, only that one should be 0, not others, which is not the case with this implementation I think, right?

dsandeep0138 replied:
Shouldn't this  be still there to handle _MultimetricScorer() failure? Please check the latest patch and let me know

thomasjpfan replied:
We can have the whole logic handled in `_score`:

```py
try:
    if y_test is None:
        scores = scorer(estimator, X_test)
    else:
        scores = scorer(estimator, X_test, y_test)
except:
	if error_score == 'raise':
		...
```

adrinjalali commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -331,26 +331,61 @@ def test_cross_validate_invalid_scoring_param():
+    with pytest.warns(UserWarning, match=warning_message) as record:
+        cross_validate(estimator, X, y, scoring=multiclass_scorer)
+    assert len(record) > 0
+    for item in record:
+        assert 'Traceback (most recent call last):\n' in str(item.message)
+        split = str(item.message).splitlines()
+        assert error_message in split[-1]
```
same here, could be removed

adrinjalali commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -331,26 +331,61 @@ def test_cross_validate_invalid_scoring_param():
+    with pytest.warns(UserWarning, match=warning_message) as record:
+        cross_validate(estimator, X, y, scoring={"foo": multiclass_scorer})
+    assert len(record) > 0
+    for item in record:
+        assert 'Traceback (most recent call last):\n' in str(item.message)
+        split = str(item.message).splitlines()
+        assert error_message in split[-1]
```
and here

adrinjalali commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -331,26 +331,61 @@ def test_cross_validate_invalid_scoring_param():
+                       "partition for these parameters will be set to %f. "
+                       "Details: \n" % score_kwargs['error_score'])
+
+    with pytest.warns(UserWarning, match=warning_message) as record:
```
and you don't need to record it

adrinjalali commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1703,6 +1752,47 @@ def test_fit_and_score_working():
+        assert error_message in split[-1]
+
+    # Test if an exception is raised on error_score='raise'
+    assert_raises_regex(ValueError, ("Classification metrics can't handle "
```
you could do `with pytest.raises(...): ...`

dsandeep0138 committed:
Update score only for the scorer that fails.

Also handle the case, when MultiMetric scorer fails and throws
an exception.

dsandeep0138 committed:
Minor fix change

dsandeep0138 committed:
Fix lint issues

thomasjpfan reviewed (commented):
Thank you for the PR @dsandeep0138 !

dsandeep0138 commented on sklearn/model_selection/_validation.py:
```diff
@@ -573,13 +573,33 @@ def _fit_and_score(estimator, X, y, scorer, train, test, verbose,
+                              (error_score, format_exc()),
+                              UserWarning)
+            else:
+                raise ValueError("error_score must be the string 'raise' or a "
+                                 "numeric value. (Hint: if using 'raise', "
+                                 "please make sure that it has been "
+                                 "spelled correctly.)")
```
we can't check this in the beginning of _fit_and_score right? Because we should be checking error_score only when fit or score failed.. we cannot check it at the beginning when we don't know if the fit/score actually fails or passes.

thomasjpfan replied:
Validating the input `error_score` in the beginning is not too costly and allows the rest of the code to not constantly need to check:

```py
if not (isinstance(error_score, numbers.Number) or error_score == 'raise'):
	raise ValueError("error_score must be a string 'raise' or ...")

progress_msg = ""
...
```

dsandeep0138 replied:
Lets say error_score is 'raise', but fit and score function doesn't fail. Then we incorrectly raise the ValueError if we add it to the beginning of the function and check it everytime. We should check only when fit/score fails

dsandeep0138 replied:
The thing is: if fit and score are passing, then irrespective of error_score, we have to return the scores received. But if we check error_score at the beginning, we deviate from that behavior.

thomasjpfan replied:
> Lets say error_score is 'raise', but fit and score function doesn't fail.

If `error_score` is `'raise`' then the above condition will not raise. It only raise if `error_score` is not a number and its not equal to `'raise'`.

> The thing is: if fit and score are passing, then irrespective of error_score, we have to return the scores received. But if we check error_score at the beginning, we deviate from that behavior.

In our public api documentation for [cross_validate_score](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html#sklearn.model_selection.cross_val_score) or [GridSearchCV](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV) it states that `error_score` must be numeric or `'raise'`. If a user passes an invalid `error_score` I think we should have already raised an error without doing any calculation.

thomasjpfan replied:
> The thing is: if fit and score are passing, then irrespective of error_score, we have to return the scores received. But if we check error_score at the beginning, we deviate from that behavior.

This is a valid concern. For this specific case, I think doing input validation first will be better long run. Let's see a user provides `error_score='wow'`, I would want that to fail right away. Otherwise a user would think that `error_score='wow'` is a valid input which is not what the API docs states.

dsandeep0138 replied:
Updated. Please review

dsandeep0138 committed:
Make fit function fail if error_score is invalid.

dsandeep0138 committed:
Merge branch 'master' into issue#17589

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -633,15 +654,34 @@ def _score(estimator, X_test, y_test, scorer):
+                    raise ValueError(error_msg % (score, type(score), name))
+                elif isinstance(error_score, numbers.Number):
+                    scores[name] = error_score
+                    warnings.warn("Scoring failed. The score on this "
```
I don't think we should replace the current error message... This is a problem of the API of the scorer, not its success at scoring.

dsandeep0138 replied:
Hi @jnothman , do you mean we should still raise ValueError here when scoring failed irrespective of error_score (I mean, we shouldn't change score to error_score when scoring failed?)

Do you suggest we should handle the case only when MultiMetricScorer throws an exception?

jnothman replied:
Yes, I would be inclined to still raise an error in this case, as closely the score is invalid, not the score

dsandeep0138 replied:
ok, let me update it to change score to error_score only when MultiMetricScorer throws an exception

dsandeep0138 replied:
Please review the latest patch.

dsandeep0138 committed:
Update to error_score only when the scorer fails.

dsandeep0138 committed:
Merge branch 'issue#17589' of https://github.com/dsandeep0138/scikit-learn into issue#17589

dsandeep0138 committed:
Fix issues with the test case.

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -500,6 +500,11 @@ def _fit_and_score(estimator, X, y, scorer, train, test, verbose,
         estimator : estimator object
             The fitted estimator.
     """
+    if not (isinstance(error_score, numbers.Number) or error_score == 'raise'):
```
Identical logic, improved legibility:
```suggestion
    if not isinstance(error_score, numbers.Number) and error_score != 'raise':
```

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -610,19 +613,37 @@ def _fit_and_score(estimator, X, y, scorer, train, test, verbose,
     Will return a dict of floats if `scorer` is a dict, otherwise a single
     float is returned.
     """
+    if not (isinstance(error_score, numbers.Number) or error_score == 'raise'):
```
```suggestion
    if not isinstance(error_score, numbers.Number and error_score != 'raise':
```

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -610,19 +613,37 @@ def _fit_and_score(estimator, X, y, scorer, train, test, verbose,
+    except Exception:
+        if error_score == 'raise':
+            raise
+        elif isinstance(error_score, numbers.Number):
```
The type has been checked.
```suggestion
        else:
```

jnothman commented on sklearn/model_selection/_validation.py:
```diff
@@ -610,19 +613,37 @@ def _fit_and_score(estimator, X, y, scorer, train, test, verbose,
 
 
-def _score(estimator, X_test, y_test, scorer):
+def _score(estimator, X_test, y_test, scorer, error_score=np.nan):
```
Giving this a default value means that we risk missing usages that do not specify the behaviour. Leave it as
```suggestion
def _score(estimator, X_test, y_test, scorer, error_score):
```

jnothman reviewed (commented):
Thanks

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1720,6 +1721,38 @@ def test_fit_and_score_working():
+    iris = load_iris()
+    X, y = iris.data, iris.target
+    clf = MockClassifier()
+    score_kwargs = {'error_score': np.nan}
```
I think it's strange that this is only used to format a warning message, and then only one key of it. Remove this variabe.

dsandeep0138 committed:
Remove default value of error_score in _score function

dsandeep0138 committed:
Add error_score argument to RFE fit.

thomasjpfan commented on sklearn/feature_selection/_rfe.py:
```diff
@@ -544,6 +545,11 @@ def fit(self, X, y, groups=None):
             train/test set. Only used in conjunction with a "Group" :term:`cv`
             instance (e.g., :class:`~sklearn.model_selection.GroupKFold`).
 
+        error_score : 'raise' or numeric, default=np.nan
+            Value to assign to the score if an error occurs in scoring
+            fitting. If set to 'raise', the error is raised.
+            If a numeric value is given, UserWarning is raised.
```
Adding this feature is outside the scope of the PR. For the call to `_score` in `RFE`, lets set `error_score='raise'` which is the behavior on master.

dsandeep0138 committed:
Remove error_score from RFE fit.

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1720,6 +1719,37 @@ def test_fit_and_score_working():
+                     " numeric value. (Hint: if using 'raise', please"
+                     " make sure that it has been spelled correctly.)")
+
+    # Test if an exception is raised when error_score is invalid string
```
we test this elsewhere. please remove it

jnothman commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -1720,6 +1719,37 @@ def test_fit_and_score_working():
+    # the warning message we're expecting to see
+    warning_message = ("Scoring failed. The score on this train-test "
+                       "partition for these parameters will be set to %f. "
+                       "Details: \n" % np.nan)
```
we should run this with two different values of error_score to make sure that it does what it's meant to

jnothman reviewed (commented):
The test case added does not isolate the condition we are trying to test: the behaviour when the scorer fails on some, but not all, runs.

Instead of the test relying on MockClassifier to give us garbage output, we should be implementing a scorer that gives valid scores sometimes but not others (e.g. failing every second time).

glemaitre commented:
@dsandeep0138 Would you be able to complete this Pull Request by implementing the test as mentioned by @jnothman
